Hotfix/fix dcs bwl issues

kantera800 opened the pull request:
all DCS code has been ported #562

DCS Validation has been started on both RDKB/UGW platforms.

This PR is used to fix all bwl related issues that has been spotted during the validation process.

adam1985d committed:
mon_wlan_dwpal: added received result size to error log print

Print dwpal_nl_cmd_get returned value, containing the received
result size if the latter does not match the expected result size.

Signed-off-by: Adam Dov <adamx.dov@intel.com>

adam1985d committed:
process_dwpal_nl_event: interface name in event is optional

Before this change, if an event did not contain an interface name,
the function returned an error. This return must be removed since
there are NL events that do not contain interface name
(e.g. "Channel_Scan_Finished" event).
After this change, failure to get the interface name from the NL event
in process_dwpal_nl_event will not return an error.
Additionally, changed log print from error to warning.

Signed-off-by: Adam Dov <adamx.dov@intel.com>

adam1985d committed:
base_wlan_hal_dwpal:fsm_setup fix wrong NL pointer validation

When an attach event is received, the dwpal_driver_nl_fd_get is not performed
due to wrong condition check of the m_dwpal_nl_ctx pointer.
Fix the wrong condition check so netlink attach can be performed.

Signed-off-by: Adam Dov <adamx.dov@intel.com>

adam1985d committed:
mon_wlan_hal: set sequence on first scan results event

When a triggered scan completes, an empty 'Channel_Scan_Dump_Result' event
sent to signal the scan is completed and that results are ready.
At this point, if dump-results is requested, it will result with a
series of 'Channel_Scan_Dump_Result' events with the results.
Fix current logic by treating the first 'Channel_Scan_Dump_Result' as results-ready
and trigger the dump-result sequence.

Signed-off-by: Adam Dov <adamx.dov@intel.com>

Commit:
bwl: dwpal: fix wrong buffer initialization

Default initialization of the "smart" results_buffer is wrong in this scenario.
results_buffer is actually a shared pointer to a uint8_t buffer.
so, calling a shared_ptr default constructor is actually writing zeros
to a buffer of an unknown size.

Since results structure also contains non basic types (std::vector)
we need to create shared pointer with instance of the
actual results structure (not using smart buffer anymore).

Finally the initialization will be done in the definition of the structure.

Signed-off-by: alex kanter <alexx.kanter@intel.com>

Commit:
bwl: dwpal: fix nl msg get

Since we're expecting a Solicited (asynchronous) event from the driver,
and it's impossible to know the type of the received message without
processing it (using a callback function), we continue processing events
until data_size != 0.
also, added 1 sec timeout is added so we dont get stuck here forever.
In practice, the response usually arrives much faster.

Signed-off-by: alex kanter <alexx.kanter@intel.com>

Commit:
bwl: dwpal: fix bssid get in nl data parsing

Fixing wrong usage of std::copy_n.
syntax of copy_n ( InputIt first, Size count, OutputIt result )
InputIt and OutputIt were switched and as the result bssid was wrongly copied to results buffer.

Signed-off-by: alex kanter <alexx.kanter@intel.com>

vitalybu commented on common/beerocks/bwl/dwpal/mon_wlan_hal_dwpal.h:
```diff
@@ -59,12 +59,16 @@ class mon_wlan_hal_dwpal : public base_wlan_hal_dwpal, public mon_wlan_hal {
                                                      LTQ_NL80211_VENDOR_SUBCMD_GET_SCAN_PARAMS,
-                                                     m_nl_buffer, NL_MAX_REPLY_BUFFSIZE)) {
+                                                     m_nl_buffer, NL_MAX_REPLY_BUFFSIZE))) {
             LOG(ERROR) << "LTQ_NL80211_VENDOR_SUBCMD_GET_SCAN_PARAMS failed! returned size does "
```
No need to specify internal structure name in the error message.
A more appropriate error message would be:
```
LOG(ERROR) << "LTQ_NL80211_VENDOR_SUBCMD_GET_SCAN_PARAMS failed: expected size" 
           << expected_result_size << ", received size " << received_result_size;
```

vitalybu commented on common/beerocks/bwl/dwpal/mon_wlan_hal_dwpal.cpp:
```diff
@@ -1058,8 +1058,7 @@ bool mon_wlan_hal_dwpal::process_dwpal_nl_event(struct nl_msg *msg)
         (if_indextoname(nla_get_u32(tb[NL80211_ATTR_IFINDEX]), ifname) == NULL)) {
-        LOG(ERROR) << __func__ << " failed to get ifname";
-        return false;
+        LOG(WARNING) << __func__ << " failed to get ifname for nl_event=" << int(gnlh->cmd);
```
There's no need for the `__func__` macro. 
The log messages already contain the filename and line number.
```suggestion
        LOG(WARNING) << "Failed to get ifname for nl_event = " << int(gnlh->cmd);
```

vitalybu commented on common/beerocks/bwl/dwpal/base_wlan_hal_dwpal.cpp:
```diff
@@ -146,7 +146,8 @@ bool base_wlan_hal_dwpal::fsm_setup()
                         LOG(DEBUG)
-                            << "dwpal_driver_nl_attach() success for: " << m_radio_info.iface_name;
+                            << "dwpal_driver_nl_attach() success for: " << m_radio_info.iface_name
+                            << ", nl_context=" << m_dwpal_nl_ctx;
```
```suggestion
                            << ", nl_context = " << m_dwpal_nl_ctx;
```

vitalybu commented on common/beerocks/bwl/dwpal/mon_wlan_hal_dwpal.cpp:
```diff
@@ -1111,7 +1111,7 @@ bool mon_wlan_hal_dwpal::process_dwpal_nl_event(struct nl_msg *msg)
             }
             // Initialize the message
-            results_buff = {};
+            memset(results_buff.get(), 0, sizeof(sCHANNEL_SCAN_RESULTS_NOTIFICATION));
```
Are you sure that this `memset` doesn't cause a build error with UGW's GCC 8.x?

tomereli replied:
Indeed, please make sure to build this PR in UGW with GCC8.x.
Resolve this if it passes.

kantera800 replied:
you have 40 cases of memset in the code currently so yes pretty sure.

vitalybu commented on common/beerocks/bwl/dwpal/base_wlan_hal_dwpal.cpp:
```diff
@@ -658,14 +658,20 @@ size_t base_wlan_hal_dwpal::dwpal_nl_cmd_get(const std::string &ifname, unsigned
-        LOG(ERROR) << " dwpal_driver_nl_msg_get failed,"
-                   << " ctx=" << m_dwpal_nl_ctx;
-        return data_size;
+    // Reply from driver could be delayed so we need
```
```suggestion
    // The reply from driver could be delayed or reordered so we need
```

tomereli replied:
Commit message - "Set nel message- typo bug fixed" - Looks like a commit message typo.

vitalybu commented on common/beerocks/bwl/dwpal/base_wlan_hal_dwpal.cpp:
```diff
@@ -658,14 +658,20 @@ size_t base_wlan_hal_dwpal::dwpal_nl_cmd_get(const std::string &ifname, unsigned
-                   << " ctx=" << m_dwpal_nl_ctx;
-        return data_size;
+    // Reply from driver could be delayed so we need
+    // to continue polling the answer until with valid size.
```
```suggestion
    // to continue polling until a response with a valid size is received.
```

vitalybu commented on common/beerocks/bwl/dwpal/base_wlan_hal_dwpal.cpp:
```diff
@@ -658,14 +658,20 @@ size_t base_wlan_hal_dwpal::dwpal_nl_cmd_get(const std::string &ifname, unsigned
+            return data_size;
+        }
     }
-
```
I actually prefer separating logical code blocks with a new line...

vitalybu commented on common/beerocks/bwl/dwpal/base_wlan_hal_dwpal.cpp:
```diff
@@ -658,14 +658,20 @@ size_t base_wlan_hal_dwpal::dwpal_nl_cmd_get(const std::string &ifname, unsigned
+    auto timeout = std::chrono::steady_clock::now() + std::chrono::seconds(1);
+    while ((data_size == 0) && std::chrono::steady_clock::now() < timeout) {
+        //parsing will be done in callback func
+        if (dwpal_driver_nl_msg_get(m_dwpal_nl_ctx, DWPAL_NL_SOLICITED_EVENT, nl_handler_cb,
```
Since `dwpal_driver_nl_msg_get` is blocking, theoretically it can block the calling thread for more than 1 second.
In practice, the response usually arrives much faster. Perhaps the comment should be updated.

tomereli replied:
@vitalybu 's explanation is excellent - please use it in the comment.

kantera800 replied:
clearly that is not what i meant you to understand from the comment,

dwpal_nl_cmd_get is blocking , inside i run   dwpal_driver_nl_msg_get in a loop with 1 sec TO 
I'll change the comment....

tomereli commented on common/beerocks/bwl/dwpal/mon_wlan_hal_dwpal.h:
```diff
@@ -59,12 +59,16 @@ class mon_wlan_hal_dwpal : public base_wlan_hal_dwpal, public mon_wlan_hal {
+        size_t received_result_size       = 0;
 
-        if (expected_result_size != dwpal_nl_cmd_get(m_radio_info.iface_name,
+        if (expected_result_size !=
```
You've got to be kidding.

tomereli replied:
If the previous comment is not clear - please `received_result_size` above instead of setting it to 0.
Then change the below to something that normal people can read:
```
if (received_result_size != expected_result_size ) {
   ...
}
```

tomereli replied:
I don't see this is changed

kantera800 replied:
didn't give a chance to push. please wait for me to re-trigger re-review next time .

vitalybu commented on common/beerocks/bwl/dwpal/base_wlan_hal_dwpal.cpp:
```diff
@@ -658,14 +658,20 @@ size_t base_wlan_hal_dwpal::dwpal_nl_cmd_get(const std::string &ifname, unsigned
+    // This is a blocking function so we don't want to stay here
+    // more than 1 sec.
+    auto timeout = std::chrono::steady_clock::now() + std::chrono::seconds(1);
+    while ((data_size == 0) && std::chrono::steady_clock::now() < timeout) {
```
The `(data_size == 0)` condition should be explained.
Since we're expecting a Solicited (asynchronous) event from the driver, and it's impossible to know the type of the received message without processing it (using a callback function), we continue processing events until `data_size != 0`.
The `data_size` is set from the processing callback only after processing the expected message.

tomereli commented on common/beerocks/bwl/dwpal/mon_wlan_hal_dwpal.h:
```diff
@@ -75,12 +79,16 @@ class mon_wlan_hal_dwpal : public base_wlan_hal_dwpal, public mon_wlan_hal {
     bool dwpal_get_scan_params_bg(sScanCfgParamsBG &params)
     {
         const size_t expected_result_size = sizeof(params) + NL_ATTR_HDR;
+        size_t received_result_size       = 0;
```
Same here, obviously.

tomereli commented on common/beerocks/bwl/dwpal/mon_wlan_hal_dwpal.cpp:
```diff
@@ -1080,26 +1080,28 @@ bool mon_wlan_hal_dwpal::process_dwpal_nl_event(struct nl_msg *msg)
         // 1st event "empty dump result" need to send back scan dump request.
         // rest of events are the actual scan dumps that need to be parsed.
         if (m_nl_seq == 0) {
-            LOG(DEBUG) << "Results dump are ready with sequence number: " << (int)nlh->nlmsg_seq;
```
The commit message is really not clear and has lots of grammar mistakes. Please fix it. Also please include references to where you get this "the nlh->nlmsg_seq is 0 for "scan_results_ready" msg" statement.

tomereli commented on common/beerocks/bwl/dwpal/mon_wlan_hal_dwpal.h:
```diff
@@ -98,8 +98,8 @@ class mon_wlan_hal_dwpal : public base_wlan_hal_dwpal, public mon_wlan_hal {
     {
-        if (dwpal_nl_cmd_set(m_radio_info.iface_name, LTQ_NL80211_VENDOR_SUBCMD_SET_SCAN_PARAMS,
-                             (unsigned char *)&params, sizeof(params))) {
+        if (!dwpal_nl_cmd_set(m_radio_info.iface_name, LTQ_NL80211_VENDOR_SUBCMD_SET_SCAN_PARAMS,
```
This looks like it should be a separate commit or at least explained in the commit message.

tomereli commented on common/beerocks/bwl/dwpal/mon_wlan_hal_dwpal.h:
```diff
@@ -98,8 +98,8 @@ class mon_wlan_hal_dwpal : public base_wlan_hal_dwpal, public mon_wlan_hal {
-        if (dwpal_nl_cmd_set(m_radio_info.iface_name, LTQ_NL80211_VENDOR_SUBCMD_SET_SCAN_PARAMS,
-                             (unsigned char *)&params, sizeof(params))) {
+        if (!dwpal_nl_cmd_set(m_radio_info.iface_name, LTQ_NL80211_VENDOR_SUBCMD_SET_SCAN_PARAMS,
+                              (unsigned char *)&params, sizeof(sScanCfgParams))) {
```
Please, no C style casts.

tomereli commented on common/beerocks/bwl/dwpal/mon_wlan_hal_dwpal.h:
```diff
@@ -108,8 +108,8 @@ class mon_wlan_hal_dwpal : public base_wlan_hal_dwpal, public mon_wlan_hal {
     {
-        if (dwpal_nl_cmd_set(m_radio_info.iface_name, LTQ_NL80211_VENDOR_SUBCMD_SET_SCAN_PARAMS_BG,
-                             (unsigned char *)&params, sizeof(params))) {
+        if (!dwpal_nl_cmd_set(m_radio_info.iface_name, LTQ_NL80211_VENDOR_SUBCMD_SET_SCAN_PARAMS_BG,
```
Same here.

tomereli commented on common/beerocks/bwl/dwpal/mon_wlan_hal_dwpal.cpp:
```diff
@@ -361,8 +361,8 @@ static bool translate_nl_data_to_bwl_results(sChannelScanResults &results,
                                              const struct nlattr **bss)
 {
     //get bssid
-    std::copy_n(results.bssid.oct, sizeof(results.bssid.oct),
```
Commit message typo - "bssid was wrongly copyied to results buffer." --> "bssid was wrongly copied to results buffer."

tomereli commented on common/beerocks/bwl/dwpal/mon_wlan_hal_dwpal.cpp:
```diff
@@ -361,8 +361,8 @@ static bool translate_nl_data_to_bwl_results(sChannelScanResults &results,
     //get bssid
-    std::copy_n(results.bssid.oct, sizeof(results.bssid.oct),
-                (unsigned char *)nla_data(bss[NL80211_BSS_BSSID]));
+    std::copy_n((unsigned char *)nla_data(bss[NL80211_BSS_BSSID]), sizeof(results.bssid),
```
Please don't use C style casts.

Commit:
bwl: dwpal: fix nl msg set

Wrong error condition validation used for dwpal_nl_cmd_set().
Also, dwpal_nl_cmd_set() proxy "vendor data" without modifying it
with unnecessary C style casting.

Fix error condition.
Change dwpal_nl_cmd_set signature to accept (const void *) and
cast vendor data (cpp style) inside before passing it to dwpal lib.

Signed-off-by: Alex Kanter <alexx.kanter@intel.com>

tomereli commented on common/beerocks/bwl/dwpal/mon_wlan_hal_dwpal.cpp:
```diff
@@ -1058,8 +1058,7 @@ bool mon_wlan_hal_dwpal::process_dwpal_nl_event(struct nl_msg *msg)
         (if_indextoname(nla_get_u32(tb[NL80211_ATTR_IFINDEX]), ifname) == NULL)) {
-        LOG(ERROR) << __func__ << " failed to get ifname";
-        return false;
+        LOG(WARNING) << "Failed to get ifname for nl_event = " << int(gnlh->cmd);
```
So why is it a warning? should be simply:
```
        if (tb[NL80211_ATTR_IFINDEX]) {
            if_indextoname(nla_get_u32(tb[NL80211_ATTR_IFINDEX]), ifname);
        }
```

vitalybu replied:
Because `if_indextoname` can also fail?
But since we continue anyway, @tomereli is probably right. No need to print anything here.

kantera800 replied:
removed the warning.
this is how it was originally but who remembers :)

tomereli commented on common/beerocks/bwl/dwpal/base_wlan_hal_dwpal.cpp:
```diff
@@ -146,7 +146,8 @@ bool base_wlan_hal_dwpal::fsm_setup()
                         m_dwpal_nl_ctx = nullptr;
                     } else {
                         LOG(DEBUG)
-                            << "dwpal_driver_nl_attach() success for: " << m_radio_info.iface_name;
```
I really dislike the "before this change - after this change" commit messages. 
I propose to change the commit message to something like this:

> When an attach event is received, the dwpal_driver_nl_fd_get is not performed due to wrong condition check of the m_dwpal_nl_ctx pointer.
Fix the wrong condition check.

tomereli commented on common/beerocks/bwl/dwpal/mon_wlan_hal_dwpal.cpp:
```diff
@@ -1080,26 +1080,29 @@ bool mon_wlan_hal_dwpal::process_dwpal_nl_event(struct nl_msg *msg)
     case Event::Channel_Scan_Dump_Result: {
```
Commit message is too long and not clear.
Suggestion:
>When a triggered scan completes, an empty 'Channel_Scan_Dump_Result' event
sent to signal the scan is completed and that results are ready.
At this point, if dump-results is requested, it will result with a
series of 'Channel_Scan_Dump_Result' events with the results.
Fix current logic by treating the first 'Channel_Scan_Dump_Result' as results-ready
and trigger the dump-result sequence.

tomereli commented on common/beerocks/bwl/dwpal/mon_wlan_hal_dwpal.cpp:
```diff
@@ -1112,7 +1112,7 @@ bool mon_wlan_hal_dwpal::process_dwpal_nl_event(struct nl_msg *msg)
             }
             // Initialize the message
-            results_buff = {};
+            memset(results_buff.get(), 0, sizeof(sCHANNEL_SCAN_RESULTS_NOTIFICATION));
```
Now this is a good commit message!

tomereli commented on common/beerocks/bwl/dwpal/base_wlan_hal_dwpal.cpp:
```diff
@@ -658,12 +658,21 @@ size_t base_wlan_hal_dwpal::dwpal_nl_cmd_get(const std::string &ifname, unsigned
-        LOG(ERROR) << " dwpal_driver_nl_msg_get failed,"
-                   << " ctx=" << m_dwpal_nl_ctx;
-        return data_size;
+    // Since we're expecting a Solicited (asynchronous) event from the driver,
```
Commit message:
>also, added 1 sec TO is added so we dont get stuck here forever.
In practice, the response usually arrives much faster.

Should change to

>also, added 1-second timeout is added so we don't get stuck here forever.
In practice, the response usually arrives much faster.

tomereli reviewed (approved):
Mainly nitpicks about the commit messages so approving.
But fix the commit messages before merging please!
Use proper English - no more "Before this change" - "After this change" formats.
Commit message should be like a story - you explain to the reader what was the issue and how it is solved by this commit. Examples inline.

tomereli commented on common/beerocks/bwl/dwpal/mon_wlan_hal_dwpal.h:
```diff
@@ -90,8 +92,8 @@ class mon_wlan_hal_dwpal : public base_wlan_hal_dwpal, public mon_wlan_hal {
-        if (dwpal_nl_cmd_set(m_radio_info.iface_name, LTQ_NL80211_VENDOR_SUBCMD_SET_SCAN_PARAMS,
-                             (unsigned char *)&params, sizeof(params))) {
+        if (!dwpal_nl_cmd_set(m_radio_info.iface_name, LTQ_NL80211_VENDOR_SUBCMD_SET_SCAN_PARAMS,
+                              reinterpret_cast<unsigned char *>(&params), sizeof(params))) {
```
This is the reason this fails to build on RAX40 - I'm now very glad I changed it to GCC8.3.
This is also the reason we never want to use C style cast and instead use C++ style casts.
The issue here is that this cast removes constness of the params argument. This means that the scan params should not be const, or that `dwpal_nl_cmd_set` signature should be modified to accept const.

vitalybu replied:
Current signature for `dwpal_nl_cmd_set` is:
```C++
bool base_wlan_hal_dwpal::dwpal_nl_cmd_set(const std::string &ifname, unsigned int nl_cmd,
                                           unsigned char *vendor_data, size_t vendor_data_size)
```

Since this function proxy the so-called "vendor data" without modifying it, it should be changed to accept `const void*`. This will remove the need for casting at all.

```C++
bool base_wlan_hal_dwpal::dwpal_nl_cmd_set(const std::string &ifname, unsigned int nl_cmd,
                                           const void *vendor_data, size_t vendor_data_size)
```

kantera800 replied:
Although changing "vendor_data" signature to (const void *) seems like the good solution i still prefer to pass it as (const unsigned char *) and send  to the dwpal library a non-const copy of it.
The reason is that @vitalybu solution still requires explicitly casting to **(const void *)** and back inside dwpal_nl_cmd_set (or leaving it as implicit cast for the compiler's decision) .
I believe casting to void * is less safer.

vitalybu commented on common/beerocks/bwl/dwpal/mon_wlan_hal_dwpal.h:
```diff
@@ -59,12 +59,13 @@ class mon_wlan_hal_dwpal : public base_wlan_hal_dwpal, public mon_wlan_hal {
-                          "not match sScanCfgParams!";
+        if (expected_result_size != received_result_size) {
+            LOG(ERROR) << "LTQ_NL80211_VENDOR_SUBCMD_GET_SCAN_PARAMS failed! expected size = "
+                       << expected_result_size << " received size = " << received_result_size;
```
```suggestion
                       << expected_result_size << ", received size = " << received_result_size;
```

vitalybu commented on common/beerocks/bwl/dwpal/mon_wlan_hal_dwpal.h:
```diff
@@ -75,12 +76,13 @@ class mon_wlan_hal_dwpal : public base_wlan_hal_dwpal, public mon_wlan_hal {
-                          "not match sScanCfgParamsBG!";
+        if (expected_result_size != received_result_size) {
+            LOG(ERROR) << "LTQ_NL80211_VENDOR_SUBCMD_GET_SCAN_PARAMS_BG failed! expected size = "
+                       << expected_result_size << " received size = " << received_result_size;
```
```suggestion
                       << expected_result_size << ", received size = " << received_result_size;
```

vitalybu commented on common/beerocks/bwl/dwpal/mon_wlan_hal_dwpal.cpp:
```diff
@@ -1080,26 +1080,29 @@ bool mon_wlan_hal_dwpal::process_dwpal_nl_event(struct nl_msg *msg)
-        } else if (m_nl_seq == nlh->nlmsg_seq) {
-            LOG(DEBUG) << "DWPAL NL event channel scan results dump";
+        if (m_nl_seq == nlh->nlmsg_seq) {
+            LOG(DEBUG) << "DWPAL NL event channel scan results dump, seq=" << int(nlh->nlmsg_seq);
```
```suggestion
            LOG(DEBUG) << "DWPAL NL event channel scan results dump, seq = " << int(nlh->nlmsg_seq);
```

vitalybu commented on common/beerocks/bwl/dwpal/mon_wlan_hal_dwpal.cpp:
```diff
@@ -361,8 +361,8 @@ static bool translate_nl_data_to_bwl_results(sChannelScanResults &results,
     //get bssid
-    std::copy_n(results.bssid.oct, sizeof(results.bssid.oct),
-                (unsigned char *)nla_data(bss[NL80211_BSS_BSSID]));
+    std::copy_n(reinterpret_cast<unsigned char *>(nla_data(bss[NL80211_BSS_BSSID])),
```
`bss[NL80211_BSS_BSSID]` can be NULL...

```suggestion
    if (!bss[NL80211_BSS_BSSID]) {
        LOG(ERROR) << "Invalid BSSID in the netlink message";
        return false;
    }

    std::copy_n(reinterpret_cast<unsigned char *>(nla_data(bss[NL80211_BSS_BSSID])),
```

vitalybu commented on common/beerocks/bwl/dwpal/base_wlan_hal_dwpal.cpp:
```diff
@@ -587,16 +587,22 @@ bool base_wlan_hal_dwpal::process_nl_events()
+    // Coping vendor_data to a local non-const variable in order
+    // to avoid passing a non-const reference in all upper layers.
+    unsigned char vendor_data_copy[vendor_data_size];
+    std::copy_n(vendor_data, vendor_data_size, vendor_data_copy);
```
Why? Just to avoid a cast?
Doesn't make sense.
Also, why not change the `vendor_data` to `void*`?

kantera800 replied:
vendor_data is used in upper layer and scan functionality is depended on it .
so yes , i dont want to allow it to be changed in lower levels.

vitalybu reviewed (approved):
Need to update some commit messages, but other then that, looks good

vitalybu commented on common/beerocks/bwl/include/bwl/mon_wlan_hal_types.h:
```diff
@@ -393,19 +393,19 @@ enum eChannelScanResultChannelBandwidth : uint8_t {
 typedef struct sChannelScanResults {
     //The current service set identifier in use by the neighboring WiFi SSID. The value MAY be empty for hidden SSIDs.
-    char ssid[beerocks::message::WIFI_SSID_MAX_LENGTH];
+    char ssid[beerocks::message::WIFI_SSID_MAX_LENGTH] = {'\0'};
```
```suggestion
    char ssid[beerocks::message::WIFI_SSID_MAX_LENGTH] = {0};
```

arnout replied:
I disagree. For me at least, `'\0'` tells me it's the zero-terminator of a C string, while `0` is just a zero that doesn't tell me much.
